### DISCUSSION TITLE
Required to send same amount of data with decoding and without decoding .

### DISCUSSION THREAD

**Discussion by aqildad-create** [pschatzmann/arduino-audio-tools#796] [Q&A] [answered]
Sender logic 

VolumeStream volume(i2sStream);
RingBufferStream out;
FormatConverterStream conv(out);
StreamCopy copier(conv, volume);
SBCEncoder enc;  EncodedAudioStream encStream(out, enc); 
.
loop ...-------
.
 copier.copy();
    int copied = encStream.readBytes((uint8_t *)Filedatas.charF , copyBytes);
       Filedatas.charF--------------> send to espnow
loop---------- 

At receiver end  ....
SBCDecoder dec;;EncodedAudioStream decStream(i2s, dec);
void OnDataRecv(const uint8_t * mac_addr, const uint8_t *incomingData, int len) {
 decStream.write((uint8_t *)Filedatas.charF , Filedatas.FiLEDATAsze);
..

Is some thing wrong in above technique ? 

Although its working and i am getting proper audio out at receiver . 

But my concerns is :
i can send 330samples per seconds of 250 bytes so total bytes send per seconds are
330x250 = 82,500 bytes . 

And my i2s sample rate s 44.1kHZ so from this total bytes / seconds will  be 44100x2( because 16 bits ) = 88200/ bytes per seconds . 

The difference only is 88500-88200 =6000 bytes we can say bytes send rate are approximately equal because 6k bytes difference is not huge .

So my question is why without decoding and with decoding i need to send same amount of data ?  

or else their is some other thing which i  am ignoring and i don't know it ? pls guide

**Comment by pschatzmann** [comment, marked as answer]
I don't understand your question. I would expect that the major thing to worry is that you keep the effective transmitted data rate (after decoding) below the capacity.

There is no connection between the amount of data that you write and the amount of data that is sent. This is decided by the Decoder or Encoder.

With what effective rate are you producing the data ? Please note that if you take an I2SStream at the source it will provide at the sampling rate. If you use a file or sine generator, it will be much faster!

**Reply by pschatzmann** [reply]
I don't see any bug here, so I convert this into a discussion!

**Comment by aqildad-create** [comment]
My point is my i2sstream source is @44100 @ 16 Bits so without encoder/decoder i need to process  around 88200 bytes .  am  I correct ??

Now  with reading bytes from SBC encoder  and send to espnow i must send maximum around 82,500 bytes  which is almost equal to above raw data 88200 bytes .   

so why i  need this large data to send even if i am using a sbc  encoder ,  i was expecting almost half of data to send as compare to original raw data because the encoder will compress the data .. I am  correct ?

**Comment by pschatzmann** [comment]
Yes 44100 at 16 bits_per_sample is 88200 bytes if you use 1 channel!

The encoder is compressing the audio so you can send at a lower rate. You can calculate the ratio by calling bytesUncompressed() / bytesCompressed()

I don't know where the 82,500 is coming from, but honestly I am not sure what the max transmission speed of esp-now with the updated idf version is ? Did you try the FAST_ESP_NOW_HACK

**Comment by aqildad-create** [comment]
82500 is espnow maximum data rate per seconds which  observe in my case . which  i think its very large because its almost equal to raw data which is 82000 bytes per seconds @44100 16 bit channel 1 . am i correct ? 

I was expecting almost half of data because encoder will compress the data . why  need to send almost 82500 bytes  per seconds to get proper audio at out put ?? 

bytesCompressed gives me 76 . from this 76  how to calculate the at what rate should  send data to espnow ?  

with bytescompreesd i am reading data from encoder like this 
frameSize=enc.bytesCompressed();
int copyBytes = frameSize*3;                                        //--------i uses 3 because 76x3=228  lower  then 250(espnow body size) 
  int copied = encStream.readBytes((uint8_t *)Filedatas.charF , copyBytes);
.Filedatas.charF ----------->>>> send this to espnow 
.

now how to calculate or lower the espnow data rate ??

**Reply by pschatzmann** [reply]
Look at my answers: your question does not make any sense to me!

**Reply by aqildad-create** [reply]
I did not try FAST_ESP_NOW_HACK yet ,i  will try it later but why i need to further fast the espnow data rate even if am achieving transfer rate of 82500 bytes ??? which is almost equal to  raw data  88200  bytes  @44100 16 bit channel 1 ?

I was hoping to reduce the transfer rate because i am using encoder (sbc).

**Reply by pschatzmann** [reply]
As I have written in my first answer: I would expect that the major thing to worry is that you keep the effective transmitted data rate (after decoding) below the capacity.

If you provide the data slower there is nothing you need to worry about!

**Comment by pschatzmann** [comment]
I just made some tests of my own and I was getting the following decoded rates using audio generated by a sine generator:
09:41:37.623 -> ==> Bytes per second: 118000
09:41:39.746 -> ==> Bytes per second: 119000
09:41:41.901 -> ==> Bytes per second: 119000
09:41:44.055 -> ==> Bytes per second: 119000
09:41:46.176 -> ==> Bytes per second: 120000
09:41:48.330 -> ==> Bytes per second: 118000
09:41:50.488 -> ==> Bytes per second: 119000

For me the decoding examples are working like a charm: https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-communication/esp-now/codec
